repair/repair.cc

```diff
@@ -1920,20 +1922,57 @@ future<> repair_service::do_rebuild_replace_with_repair(locator::token_metadata_
             auto& topology = erm->get_token_metadata().get_topology();
             std::unordered_map<dht::token_range, repair_neighbors> range_sources;
             auto nr_tables = get_nr_tables(db, keyspace_name);
-            rlogger.info("{}: started with keyspace={}, source_dc={}, nr_ranges={}, ignore_nodes={}", op, keyspace_name, source_dc, ranges.size() * nr_tables, ignore_nodes);
+            std::unordered_set<gms::inet_address> dc_nodes;
+            sstring ks_source_dc;
+            if (ignored_in_my_dc < 0) {
+                size_t ignored_in_my_dc = 0;
+                topology.for_each_node([&] (const locator::node* node) {
+                    if (node->dc_rack().dc != source_dc) {
+                        return;
+                    }
+                    if (ignore_nodes.contains(node->endpoint())) {
+                        ++ignored_in_my_dc;
+                    } else {
+                        dc_nodes.insert(node->endpoint());
+                    }
+                });
+            }
+            // Allow repairing in the dc only if there are enough replicas
+            if (!source_dc.empty() && !dc_nodes.empty()) {
+                switch (strat.get_type()) {
+                case locator::replication_strategy_type::network_topology: {
+                    if (ignored_in_my_dc) {
+                        // Don't take any chances if there are ignored nodes in this dc
+                        break;
+                    }
+                    auto opts = strat.get_config_options();
+                    auto rf = opts.contains(source_dc) ? std::atoi(opts[source_dc].c_str()) : 0;
+                    if (rf > 1) {
```

Comment:
This restriction is needed only in case source_dc is the dc of local node, in which case there will be no neighbors to repair. We could check neighbors.empty() below instead, and then use all neighbors from all dcs.

Reply:
I find it very odd we allow replacing a node from one dc with a node in a different dc.
I don't see how that would work with network topology replication strategy.

Regarding using the neightbors, I think it's incorrect as I explained above with regards to ensuring quorum.
Say the data center has 6 nodes, out of which 2 nodes are dead. rf is 3 (e.g. 1 replica per rack).
Neighbors in this case would consist 4 nodes IIUC, since 2 nodes are lost, with rf-3, it is possible that only a single replica will remain and we shouldn't repair from a single replica since this might break quorum consistency level guarantees.

Reply:
revised in v2